staging/src/k8s.io/client-go/tools/cache/reflector_watchlist_test.go

```diff
@@ -0,0 +1,441 @@
+			},
+			expectedWatchRequests: 2,
+			expectedRequestOptions: []metav1.ListOptions{
+				{SendInitialEvents: pointer.Bool(true), AllowWatchBookmarks: true, ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan, TimeoutSeconds: pointer.Int64(1)},
+				{AllowWatchBookmarks: true, ResourceVersion: "3", TimeoutSeconds: pointer.Int64(1)},
+			},
+			expectedStoreContent: []v1.Pod{*makePod("p1", "1"), *makePod("p2", "2"), *makePod("p3", "3"), *makePod("p4", "4")},
+		},
+		{
+			name: "expiring an established watcher results in returning an error from the reflector",
+			watchServerErrorPredicate: func() func(options metav1.ListOptions) error {
+				counter := 0
+				return func(options metav1.ListOptions) error {
+					counter++
+					if counter == 2 {
+						return apierrors.NewResourceExpired("rv already expired")
+					}
+					return nil
+				}
+			}(),
+			stopAfterWatchEvents:  3,
+			closeAfterWatchEvents: 5,
+			returnedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: makePod("p1", "1")},
+				{Type: watch.Bookmark, Object: &v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						ResourceVersion: "2",
+						Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+					},
+				}},
+				{Type: watch.Added, Object: makePod("p3", "3")},
+			},
+			expectedWatchRequests: 2,
+			expectedRequestOptions: []metav1.ListOptions{
+				{SendInitialEvents: pointer.Bool(true), AllowWatchBookmarks: true, ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan, TimeoutSeconds: pointer.Int64(1)},
+				{AllowWatchBookmarks: true, ResourceVersion: "3", TimeoutSeconds: pointer.Int64(1)},
+			},
+			expectedStoreContent: []v1.Pod{*makePod("p1", "1"), *makePod("p3", "3")},
+			expectedError:        apierrors.NewResourceExpired("rv already expired"),
+		},
+	}
+	for _, scenario := range scenarios {
+		t.Run(scenario.name, func(t *testing.T) {
+			targetListWatcher, targetStore, target, stopCh := testData()
+			if len(scenario.returnedWatchEvents) > 0 {
+				go func() {
+					for i, e := range scenario.returnedWatchEvents {
+						targetListWatcher.fakeWatcher.Action(e.Type, e.Object)
+						i = i + 1
+						if i == scenario.stopAfterWatchEvents {
+							targetListWatcher.StopStartWatcher()
+							continue
+						}
+						if i == scenario.closeAfterWatchEvents {
+							close(stopCh)
+						}
+					}
+				}()
+			}
+			targetListWatcher.watcherErrorPredicate = scenario.watchServerErrorPredicate
+			targetListWatcher.closeAfterWatchRequests = scenario.closeAfterWatchRequests
+			targetListWatcher.customListResponse = scenario.returnedPodList
+			targetListWatcher.closeAfterListRequests = scenario.closeAfterListRequests
+			if scenario.disableUseWatchList {
+				target.UseWatchList = false
+			}
+
+			err := target.ListAndWatch(stopCh)
+			if scenario.expectedError != nil && err == nil {
+				t.Fatalf("expected error %q, got nil", scenario.expectedError)
+			} else if scenario.expectedError == nil && err != nil {
```

Comment:
nit: remove "else" - if you entered the above, it will fail the test anyway and stop

same below